experimental/dds/tree2/src/test/shared-tree/sharedTree.spec.ts

```diff
@@ -743,6 +743,103 @@ describe("SharedTree", () => {
+			const sb = new SchemaBuilder({ scope: "shared tree undo tests" });
+			const schema = sb.intoSchema(sb.list(sb.list(sb.string)));
+
+			it("with the edit sequenced first", () => {
+				const provider = new TestTreeProviderLite(2);
+				const content = {
+					schema,
+					allowedSchemaModifications: AllowedUpdateType.None,
+					initialTree: [["a"], ["b"]] as any,
+				} satisfies InitializeAndSchematizeConfiguration;
+				const tree1 = provider.trees[0].schematizeInternal(content);
+				const { undoStack: undoStack1, unsubscribe: unsubscribe1 } =
+					createTestUndoRedoStacks(tree1.checkout);
+				const tree2 = provider.trees[1].schematizeInternal(content);
+				const { undoStack: undoStack2, unsubscribe: unsubscribe2 } =
+					createTestUndoRedoStacks(tree2.checkout);
+
+				provider.processMessages();
+
+				// Validate insertion
+				validateTreeContent(tree2.checkout, content);
+
+				// edit subtree
+				tree2.root[0].insertAtEnd(["c"]);
+				provider.processMessages();
+				assert.equal(tree1.root[0][0], "a");
+				assert.equal(tree1.root[0][1], "c");
+				assert.equal(tree2.root[0][0], "a");
+				assert.equal(tree2.root[0][1], "c");
+				// delete subtree
+				tree1.root.removeAt(0);
+				provider.processMessages();
+				assert.equal(tree1.root[0][0], "b");
+				assert.equal(tree2.root[0][0], "b");
+
+				undoStack2.pop()?.revert();
+				undoStack1.pop()?.revert();
+				provider.processMessages();
+
+				// check the undo happened
+				assert.equal(tree1.root[0][0], "a");
+				assert.equal(tree1.root[0][1], undefined);
+				assert.equal(tree2.root[0][0], "a");
+				assert.equal(tree2.root[0][1], undefined);
+
+				unsubscribe1();
+				unsubscribe2();
+			});
+
+			it("with the restore sequenced first", () => {
+				const provider = new TestTreeProviderLite(2);
+				const content = {
+					schema,
+					allowedSchemaModifications: AllowedUpdateType.None,
+					initialTree: [["a"], ["b"]] as any,
+				} satisfies InitializeAndSchematizeConfiguration;
+				const tree1 = provider.trees[0].schematizeInternal(content);
+				const { undoStack: undoStack1, unsubscribe: unsubscribe1 } =
+					createTestUndoRedoStacks(tree1.checkout);
+				const tree2 = provider.trees[1].schematizeInternal(content);
+				const { undoStack: undoStack2, unsubscribe: unsubscribe2 } =
+					createTestUndoRedoStacks(tree2.checkout);
+
+				provider.processMessages();
+
+				// Validate insertion
+				validateTreeContent(tree2.checkout, content);
+
+				// edit subtree
+				tree2.root[0].insertAtEnd(["c"]);
+				provider.processMessages();
+				assert.equal(tree1.root[0][0], "a");
+				assert.equal(tree1.root[0][1], "c");
+				assert.equal(tree2.root[0][0], "a");
+				assert.equal(tree2.root[0][1], "c");
+				// delete subtree
+				tree1.root.removeAt(0);
+				provider.processMessages();
+				assert.equal(tree1.root[0][0], "b");
+				assert.equal(tree2.root[0][0], "b");
+
+				undoStack1.pop()?.revert();
+				undoStack2.pop()?.revert();
```

Comment:
I'm guessing the only difference between this test and the one above is that these two lines are in reverse order.
I think it would be nice to de-duplicate the code and have a single `it(...)` block in a for loop that iterates over the two scenarios: `for (const scenario of ["restore then change", "change then restore"])`